### PR TITLE
Metrics: add the `approved_count` and `unmatched_count` metrics

### PR DESCRIPTION
**Pull request by maelvls**
The only metrics approver-policy currently comes with is controller-runtime's. Although these metrics are useful to know whether the reconciliation loops are doing well, they are less than ideal to understand problems with misconfigured CertificateRequestPolicies.

The two use-cases are:

1. I'd like to know when there are "orphan" certificate requests, i.e., certificate requests that haven't been matched to any CertificateRequestPolicies.
2. I'd like to know when there is an increase in the number of denials. That might indicate a misconfiguration.

In this PR, I propose to add two metrics:

| Metric | Labels | Note | 
| --- | --- | --- |
| `approverpolicy_certificaterequest_orphan_count` (enum gauge) | `namespace`\* | 0-values not shown | 
| `approverpolicy_certificaterequest_approved_count` (gauge) | `namespace`, `approved_status` (False, True) | 0-values not shown |

By "0-values not shown", I mean that the metric only shows a value when it has a non-0 value. Since the metric is labelled by namespace, you will only see the namespaces that have at least one orphan for the `orphan_count` metric (same goes with the `approved_count` metric). We haven't implemented the extra `orphan_count_total` and  `approved_count_total` as part of this PR because I didn't know that `count_total` was meant to circumvent this 0-value limitation.

We haven't added the label `name` for `unmatched_count` because:
-  It may end up being very big (cardinality problem) in case of
many CRs being created at the same time and if the metrics are scrapped
right before the approver has had time to set the Approved condition.
- The `name` isn't often useful, particularly with the csi-driver. It would be much more useful to have a label `owner_name` and `owner_kind`.

**Testing it yourself:**

First, install cert-manager without its built-in approver:

```bash
helm upgrade --install cert-manager jetstack/cert-manager --namespace cert-manager --create-namespace --set installCRDs=true \
  --set prometheus.enabled=true \
  --set prometheus.servicemonitor.enabled=true \
  --set prometheus.servicemonitor.labels.instance=primary \
  --set extraArgs="{--controllers=*\,-certificaterequests-approver,--feature-gates=StableCertificateRequestName=true}"
```

Then, install the Grafana Agent Operator:

```bash
helm repo add --force grafana https://grafana.github.io/helm-charts
helm upgrade --install -n grafana --create-namespace grafana-agent-operator grafana/grafana-agent-operator
```

Then, go to https://grafana.com/. Create an account, and find the credentials to the Prometheus push endpoint:

grafana-cloud-1

grafana-cloud-2

With the username and password, create the following secret:

```bash
kubectl create secret generic primary-credentials-metrics \
  -n grafana
  --from-literal=username=123456 \
  --from-literal=password=glc_... \
```

Then, use the URL to create the following:

```yaml
apiVersion: monitoring.grafana.com/v1alpha1
kind: MetricsInstance
metadata:
  name: primary
  namespace: grafana
  labels:
    agent: grafana-agent-metrics
spec:
  remoteWrite:
    - url: https://prometheus-prod-01-eu-west-0.grafana.net/api/prom/push # ✨
      basicAuth:
        username:
          name: primary-credentials-metrics
          key: username
        password:
          name: primary-credentials-metrics
          key: password
  serviceMonitorNamespaceSelector: {}
  serviceMonitorSelector:
    matchLabels:
      instance: primary
  podMonitorNamespaceSelector: {}
  podMonitorSelector:
    matchLabels:
      instance: primary
apiVersion: monitoring.grafana.com/v1alpha1
kind: GrafanaAgent
metadata:
  name: grafana-agent
  namespace: grafana
  labels:
    app: grafana-agent
spec:
  image: grafana/agent:v0.36.1
  integrations:
    selector:
      matchLabels:
        agent: grafana-agent-integrations
  logLevel: info
  serviceAccountName: grafana-agent
  metrics:
    instanceSelector:
      matchLabels:
        agent: grafana-agent-metrics
    externalLabels:
      cluster: cloud
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: grafana-agent
  namespace: grafana
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: grafana-agent
rules:
  - apiGroups:
      - ""
    resources:
      - nodes
      - nodes/proxy
      - nodes/metrics
      - services
      - endpoints
      - pods
      - events
    verbs:
      - get
      - list
      - watch
  - apiGroups:
      - networking.k8s.io
    resources:
      - ingresses
    verbs:
      - get
      - list
      - watch
  - nonResourceURLs:
      - /metrics
      - /metrics/cadvisor
    verbs:
      - get
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: grafana-agent
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: grafana-agent
subjects:
  - kind: ServiceAccount
    name: grafana-agent
    namespace: grafana
```

Then, build the approver-policy image:

```bash
# From the approver-policy repo:
make docker-push VERSION=dev DOCKER_REGISTRY=ttl.sh/mael
```

Then, install approver-policy:

```bash
helm upgrade approver-policy deploy/charts/approver-policy \
  --install \
  --wait \
  --namespace cert-manager \
  --values /dev/stdin <<EOF

image:
  tag: dev
  repository: ttl.sh/mael/approver-policy-enterprise
  pullPolicy: Always
app:
  approveSignerNames:
    - "issuers.cert-manager.io/*"
    - "clusterissuers.cert-manager.io/*"
  metrics:
    service:
      enabled: true
      servicemonitor:
        enabled: true
        interval: 10s
        scrapeTimeout: 5s
        labels:
          instance: primary
EOF
```

Check that the metrics are properly served:

```bash
kubectl get --raw /api/v1/namespaces/cert-manager/services/cert-manager-approver-policy:9402/proxy/metrics
```

Then, create a CertificateRequestPolicy just to test:

```yaml
apiVersion: policy.cert-manager.io/v1alpha1
kind: CertificateRequestPolicy
metadata:
  name: mael
spec:
  allowed:
    commonName:
      value: "*"
    dnsNames:
      values: ["*"]
  plugins: {}
  selector:
    issuerRef: {}
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: cert-manager-policy:mael
rules:
  - apiGroups: ["policy.cert-manager.io"]
    resources: ["certificaterequestpolicies"]
    verbs: ["use"]
    resourceNames: ["mael"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: cert-manager-policy:mael
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cert-manager-policy:mael
subjects:
  - kind: ServiceAccount
    name: cert-manager
    namespace: cert-manager
```

<details>
<summary>Optional: install the cAdvisor and kubelet service monitors so that you can see the CPU and memory usage of the approver-policy pod</summary>

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    instance: primary
  name: kubelet-monitor
  namespace: default
spec:
  endpoints:
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    honorLabels: true
    interval: 60s
    metricRelabelings:
    - action: keep
      regex: kubelet_cgroup_manager_duration_seconds_count|go_goroutines|kubelet_pod_start_duration_seconds_count|kubelet_runtime_operations_total|kubelet_pleg_relist_duration_seconds_bucket|volume_manager_total_volumes|kubelet_volume_stats_capacity_bytes|container_cpu_usage_seconds_total|container_network_transmit_bytes_total|kubelet_runtime_operations_errors_total|container_network_receive_bytes_total|container_memory_swap|container_network_receive_packets_total|container_cpu_cfs_periods_total|container_cpu_cfs_throttled_periods_total|kubelet_running_pod_count|node_namespace_pod_container:container_cpu_usage_seconds_total:sum_rate|container_memory_working_set_bytes|storage_operation_errors_total|kubelet_pleg_relist_duration_seconds_count|kubelet_running_pods|rest_client_request_duration_seconds_bucket|process_resident_memory_bytes|storage_operation_duration_seconds_count|kubelet_running_containers|kubelet_runtime_operations_duration_seconds_bucket|kubelet_node_config_error|kubelet_cgroup_manager_duration_seconds_bucket|kubelet_running_container_count|kubelet_volume_stats_available_bytes|kubelet_volume_stats_inodes|container_memory_rss|kubelet_pod_worker_duration_seconds_count|kubelet_node_name|kubelet_pleg_relist_interval_seconds_bucket|container_network_receive_packets_dropped_total|kubelet_pod_worker_duration_seconds_bucket|container_start_time_seconds|container_network_transmit_packets_dropped_total|process_cpu_seconds_total|storage_operation_duration_seconds_bucket|container_memory_cache|container_network_transmit_packets_total|kubelet_volume_stats_inodes_used|up|rest_client_requests_total|container_memory_usage_bytes
      sourceLabels:
      - __name__
    port: https-metrics
    relabelings:
    - sourceLabels:
      - __metrics_path__
      targetLabel: metrics_path
    - action: replace
      targetLabel: job
      replacement: integrations/kubernetes/kubelet
    scheme: https
    tlsConfig:
      insecureSkipVerify: true
  namespaceSelector:
    matchNames:
    - default
  selector:
    matchLabels:
      app.kubernetes.io/name: kubelet
---
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    instance: primary
  name: cadvisor-monitor
  namespace: default
spec:
  endpoints:
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    honorLabels: true
    honorTimestamps: false
    interval: 60s
    metricRelabelings:
    - action: keep
      regex: kubelet_cgroup_manager_duration_seconds_count|go_goroutines|kubelet_pod_start_duration_seconds_count|kubelet_runtime_operations_total|kubelet_pleg_relist_duration_seconds_bucket|volume_manager_total_volumes|kubelet_volume_stats_capacity_bytes|container_cpu_usage_seconds_total|container_network_transmit_bytes_total|kubelet_runtime_operations_errors_total|container_network_receive_bytes_total|container_memory_swap|container_network_receive_packets_total|container_cpu_cfs_periods_total|container_cpu_cfs_throttled_periods_total|kubelet_running_pod_count|node_namespace_pod_container:container_cpu_usage_seconds_total:sum_rate|container_memory_working_set_bytes|storage_operation_errors_total|kubelet_pleg_relist_duration_seconds_count|kubelet_running_pods|rest_client_request_duration_seconds_bucket|process_resident_memory_bytes|storage_operation_duration_seconds_count|kubelet_running_containers|kubelet_runtime_operations_duration_seconds_bucket|kubelet_node_config_error|kubelet_cgroup_manager_duration_seconds_bucket|kubelet_running_container_count|kubelet_volume_stats_available_bytes|kubelet_volume_stats_inodes|container_memory_rss|kubelet_pod_worker_duration_seconds_count|kubelet_node_name|kubelet_pleg_relist_interval_seconds_bucket|container_network_receive_packets_dropped_total|kubelet_pod_worker_duration_seconds_bucket|container_start_time_seconds|container_network_transmit_packets_dropped_total|process_cpu_seconds_total|storage_operation_duration_seconds_bucket|container_memory_cache|container_network_transmit_packets_total|kubelet_volume_stats_inodes_used|up|rest_client_requests_total|container_memory_usage_bytes
      sourceLabels:
      - __name__
    path: /metrics/cadvisor
    port: https-metrics
    relabelings:
    - sourceLabels:
      - __metrics_path__
      targetLabel: metrics_path
    - action: replace
      targetLabel: job
      replacement: integrations/kubernetes/cadvisor
    scheme: https
    tlsConfig:
      insecureSkipVerify: true
  namespaceSelector:
    matchNames:
    - default
  selector:
    matchLabels:
      app.kubernetes.io/name: kubelet
```

</details>

Finally, go to your Grafana Cloud instance and add the following dashboard by
URL: <https://raw.githubusercontent.com/maelvls/approver-policy/metrics/docs/examples/grafana-dashboard.json>

Visually, it could look like this:

Web capture_20-10-2023_195124_maelvls grafana net

### CODE REVIEW

File: pkg/internal/metrics/metrics.go
Comment (erikgb) [inline pkg/internal/metrics/metrics.go:105]:
```suggestion
		namespace string
		approvedStatus metav1.ConditionStatus
```

I would prefer if we keep the `ConditionStatus` type alias until the metrics are produced at the end of this function. IMO it would make the code more readable.

File: pkg/internal/metrics/metrics.go
Comment (erikgb) [inline pkg/internal/metrics/metrics.go:191]: Maybe use this utility function instead:

https://github.com/kubernetes/apimachinery/blob/a1b8da8152f9991a6691ae94bd6b6e3c34a590bd/pkg/api/meta/conditions.go#L91

File: pkg/internal/metrics/metrics.go
Comment (erikgb) [inline pkg/internal/metrics/metrics.go:190]: The name of this function seems wrong to me. You ignore the `reason` return param in all calls to it. 😉 Maybe call the function `getStatus` instead, and return just the condition status?

**Review by erikgb — CHANGES_REQUESTED** [review]
This looks useful! Thanks @maelvls! Added some initial review comments, but in general I think:

- the Grafana dashboard could be a separate follow-up PR. I am unable to review the dashboard, since I am not familiar with the syntax. So putting a LGTM would be hard.
- you should add some tests of the metrics collector. 😉

File: pkg/internal/metrics/metrics.go
Comment (maelvls) [reply pkg/internal/metrics/metrics.go:190]: Good point. I don't even think the generic aspect is worth it... I had written this for some other project that had lots of condition types, but here we only care about the CertificateRequest.status.conditions. 😅

File: pkg/internal/metrics/metrics.go
Comment (maelvls) [reply pkg/internal/metrics/metrics.go:191]: Unfortunately, we use a custom type `cmapi.CertificateRequestCondition` instead of `metav1.Condition` 😞

File: pkg/internal/metrics/metrics.go
Comment (maelvls) [reply pkg/internal/metrics/metrics.go]: Done!

**Comment by maelvls** [comment]
I have removed the Grafana dashboard. I also added tests. I could have tested the `hasSynced` behavior but I didn't since it seemed like a lot of effort 😔

**Comment by maelvls** [comment]
@erikgb I have updated the PR: I renamed "orphan" to "unmatched", and I removed the `name` label from the `unmatched_count` metric. WDYT?

For anyone reading this, I have removed the label `name` in the gauge `unmatched_count` because:
-  It may end up being very big (cardinality problem) in case of
many CRs being created at the same time and if the metrics are scrapped
right before the approver has had time to set the Approved condition.
- The `name` isn't often useful, particularly with the csi-driver. It would be much more useful to have a label `owner_name` and `owner_kind`.

**Comment by inteon** [comment]
/test pull-cert-manager-approver-policy-test
I think this is a flake

**Comment by inteon** [comment]
/approve
/lgtm
Thanks @wallrj!

**Comment by inteon** [comment]
/approve
/lgtm
Thanks @maelvls, the new metrics look very promising.
In the future, we could maybe add a metric specifically for the # CertReqs that this approver approves.

**Comment by maelvls** [comment]
About the flake, I was able to reproduce the error locally once. For context, the error is:

> Selector [It] if namespace selector doesn't match, don't approve or deny. If selector updated with match, should approve CertificateRequest

I reproduced it with:

```
export KUBEBUILDER_ASSETS=$PWD/_bin/kubebuilder/bin
export ROOTDIR=$PWD
go test ./pkg/internal/controllers/test -v -count=1 -ginkgo.focus="if namespace selector doesn't match, don't approve or deny"
```

I wasn't able to reproduce it more than once though... And I don't know how to debug this flake either. 🤔

**Comment by erikgb** [comment]
Thanks @maelvls! Tim is a fast reviewer, so I had no time to review. But LGTM!

**Pull request merged**
